### PR TITLE
add command edit

### LINKED ISSUE #407: Fixes to keep up with gcalcli
Hi, I am in the process of switching from gcalcli to khal (since vdirsyncer now supports google). I have noticed that in terms of commands gcalcli seems to be ahead in a few ways. I would like to help bridge that gap.

A few ideas (in order of priority for me):

allow agenda to have different formats (#273) including json (and for me TSV), or at the very least a pretty default, the setup I have looks like: 
screen shot 2016-04-25 at 9 26 56 am
this setup was created using awk piped from gcalcli tsv output, and recreated using some regex from khal agenda output

an upcoming command, in gcalcli this calls an external command with event that will start in the next N minutes, all I care about is having either at or agenda support a next N minutes start and then a reasonable output format so that I can parse for notifications

interactive new and edit/delete, these commands would either drop right into the new/edit or interactive or have their own flow, the point here would be to initiate from the command line. In gcalcli one can type `gcalcli edit SEARCH` which will loop through all matches and ask what you want to do (same with delete). We could easily have a loop over matching searches and ask if edit or delete is waste and or use a similar edit flow and the interactive new. The point is that some people (myself) do not want to open ikhal, then search, then edit, when a simple `khal edit SEARCH` would find the event and allow me to change one detail. (shot below)
screen shot 2016-04-25 at 9 35 50 am

As I am new to the project I am really just trying to gauge whether these changes are within what khal is trying to be. Please let me know any thoughts, I will hopefully start on this in the next few weeks.

### PR DESCRIPTION
**Pull request by youngmoney**
I've created a new interactive command called `edit`, it works completely in
the command line using just prompts, this will close #407, and I believe it solves #49.

This command accepts a search string, and then loops through each event and
asks if the user wants to edit it. If so the user chooses an aspect (eg title),
enters a new value and it is saved. events can also be deleted.

In the future future it would be good to add command line options and a batch
mode, this would enable other commands to edit events. Of course we would need
a way to be specific.

### CODE REVIEW

**Comment by youngmoney** [comment]
As a side note I would like to get this merged (if it will be) before I start
working on the interactive part of the `new` command so that I can reuse parts
of it.

**Comment by youngmoney** [comment]
actually, forget that (and sorry for so many comments) Ill actually push the
interactive additions to `new` to this PR as well hopefully this weekend. I
would like to see #406 closed within the next week or so.

**Comment by geier** [comment]
Can we have some tests for `aux.timedelta2str()` please? Otherwise this looks good.

BTW, in case you didn't know, if you install pytest-cov, you can do `py.test --cov=khal --cov-report html tests` and get a nice graphic coverage in a folder named `htmlcov`.

**Comment by youngmoney** [comment]
I just pushed the interactive new command option, there is a problem that
occurs when I try to update the new event, the assertion in khalendar.update
that event.etag exists is failing. I do not know enough to understand why.

**Comment by geier** [comment]
I'll have a look. 

Am 15. Juni 2016 23:06:21 MESZ, schrieb youngmoney <notifications@github.com>:
>I just pushed the interactive new command option, there is a problem
>that
>occurs when I try to update the new event, the assertion in
>khalendar.update
>that event.etag exists is failing. I do not know enough to understand
>why.
>
>---
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/pimutils/khal/pull/438#issuecomment-226320136

-- 
Sent from my phone. Please excuse my brevity.

**Comment by youngmoney** [comment]
Great! Thanks, it appears if you interactively add an event, and then it switched to interactive editing of the same event (at the end of the new function in controllers) it occurs when one interactively changes an attribute eg description

**Comment by geier** [comment]
The issue with updating a "new" event should be fixed with  9e560c9, you best rebase on master.

Some notes on `new -i`:
* I don't think it's a good idea to cram the interactive part into `new_from_string`, because it's not actually from string, better put it in `cli.new`
* putting invalid values into any field gives a generic error message ('critical: Error`), I'd suggest validating those values on the fly and reporting back to the user without exiting
* running `khal new` without any arguments gives less than helpful error message
* cannot give no description
* when creating an event the summary is called title (or vice versa), I think it's better to stay consistent
* should we tell the user the default timezone?

**Comment by youngmoney** [comment]
okay I have pushed rebased changes that address all of those concerns, I have added the options to accept "None" as an option to certain details in edit and interactive new, in addition I've factored out interactive new and made errors way better

**Comment by youngmoney** [comment]
any thoughts? I would like to finish this up this week if possible

File: khal/aux.py [hunk 1]
```diff
@@ -392,163 +444,85 @@ def generate_random_uid():
...
-                    except ValueError:
-                        pass
-                if until_date is None:
-                    logger.fatal("Cannot parse until date: '{}'\nPlease have a look "
-                                 "at the documentation.".format(until))
-                    raise FatalError()
-                rrule_settings['until'] = until_date
-
-            event.add('rrule', rrule_settings)
+def eventinfofstr(info_string, locale, default_timedelta=None,
```
Comment (geier) [inline khal/aux.py]: please write a docstring that describes what those parameters do and what you expect `info_string` to look like

File: khal/cli.py [hunk 2]
```diff
@@ -344,26 +342,45 @@ def new(ctx, calendar, start, end, timezone, summary, description, location, cat
         everything behind them is taken as the event's description.
         '''
         # ugly hack to change how click presents the help string
-        eventlist = [start, end, timezone, summary] + list(description)
-        eventlist = [element for element in eventlist if element is not None]
+        if not info and not interactive:
+                raise click.BadParameter(
+                    'no details provided, '
+                    'did you mean to use --interactive/-i'
+                )
```
Comment (geier) [inline khal/cli.py]: missing `?` ?

**Comment by geier** [comment]
Why do I need to type `None` instead of just entering an empty string?

**Comment by geier** [comment]
when creating a new event with `-i`, it's not clear, that `Edit? no` actually means save, not sure if it's worth the hassle though.

File: doc/source/usage.rst [hunk 3]
```diff
@@ -318,7 +318,8 @@ allows for adding new events. ``khal new`` should understand the following synta
 
 ::
 
-    khal new [-a CALENDAR] [OPTIONS] startdatetime [enddatetime] [timezone] summary [description]
+    khal new [-a CALENDAR] [OPTIONS] [START] [END | DELTA] [TIMEZONE] [SUMMARY]
+    [:: DESCRIPTION]
```
Comment (geier) [inline doc/source/usage.rst]: I think `khal new [-a CALENDAR] [OPTIONS] [START [END | DELTA] .... summary ... ]` would be clearer, as the order is still important.

**Comment by youngmoney** [comment]
I've made all of those changes. the reason for the None is that the prompt is handled by click and click does not accept empty prompt responses because there is an option for a default value (which is used a lot in edit and in new) and when there is a default value then hitting enter with no text is considered accepting the default value. As such the only way to get an empty string in most cases is to explicitly type  None, otherwise we could not have a default value, which would be annoying in most cases. None seems to me to be the best option, of course the wording could change to be for instance the literal empty quotes "" instead of None

**Comment by geier** [comment]
Thanks! I'll guess someone else can take care of this.

One more thing though: this currently breaks creating new events from ikhal, please have a look.

**Comment by geier** [comment]
Thanks and merged.

**Pull request merged**